haskell/src/Monocle/Api/Server.hs

```diff
@@ -299,6 +311,120 @@ crawlerCommitInfo request = do
         . Enumerated
         $ Right err
 
+-- | /task_data endpoints
+taskDataTaskDataAdd :: TaskDataPB.TaskDataAddRequest -> AppM TaskDataPB.TaskDataAddResponse
+taskDataTaskDataAdd TaskDataPB.TaskDataAddRequest {..} = do
+  tenants <- getConfig
+  let requestE = do
+        index <-
+          Config.lookupTenant tenants (toStrict taskDataAddRequestIndex)
+            `orDie` TaskDataPB.TaskDataCommitErrorUnknownApiKey
+
+        void $
+          Config.lookupCrawler index (toStrict taskDataAddRequestCrawler)
+            `orDie` TaskDataPB.TaskDataCommitErrorUnknownCrawler
+
+        when
+          (Config.crawlers_api_key index /= Just (toStrict taskDataAddRequestApikey))
+          (Left TaskDataPB.TaskDataCommitErrorUnknownApiKey)
+
+        pure index
+  case requestE of
+    Left err -> do
+      pure $
+        TaskDataPB.TaskDataAddResponse
+          . Just
+          . TaskDataPB.TaskDataAddResponseResultError
+          . Enumerated
+          $ Right err
+    Right index -> do
+      void $ runTenantM index $ do I.taskDataAdd $ toList taskDataAddRequestItems
+      pure $ TaskDataPB.TaskDataAddResponse Nothing
+
+taskDataTaskDataCommit :: TaskDataPB.TaskDataCommitRequest -> AppM TaskDataPB.TaskDataCommitResponse
+taskDataTaskDataCommit TaskDataPB.TaskDataCommitRequest {..} = do
+  tenants <- getConfig
+  let requestE = do
+        index <-
+          Config.lookupTenant tenants (toStrict taskDataCommitRequestIndex)
+            `orDie` TaskDataPB.TaskDataCommitErrorUnknownApiKey
+
+        _crawler <-
+          Config.lookupCrawler index (toStrict taskDataCommitRequestCrawler)
+            `orDie` TaskDataPB.TaskDataCommitErrorUnknownCrawler
+        when
+          (Config.crawlers_api_key index /= Just (toStrict taskDataCommitRequestApikey))
+          (Left TaskDataPB.TaskDataCommitErrorUnknownApiKey)
+
+        void taskDataCommitRequestTimestamp `orDie` TaskDataPB.TaskDataCommitErrorCommitDateInferiorThanPrevious
+
+        pure (index, _crawler)
```

Comment:
even though we don't have an error code for that issue, we could still do the lazy validation here:
```suggestion
        let ts =  T.toUTCTime $ fromMaybe (error "Missing commit timestamp in request") taskDataCommitRequestTimestamp
        pure (index, ts)
```